onecgiar-pr-server/src/api/notification/notification.service.ts

```diff
@@ -0,0 +1,449 @@
+import { NotificationDto } from '../../shared/microservices/socket-management/dto/create-socket.dto';
+import { ShareResultRequestService } from '../results/share-result-request/share-result-request.service';
+import { FindOperator, MoreThan } from 'typeorm';
+import { UserRepository } from '../../auth/modules/user/repositories/user.repository';
+
+@Injectable()
+export class NotificationService {
+  private readonly _logger = new Logger(NotificationService.name);
+
+  constructor(
+    private readonly _notificationLevelRepository: NotificationLevelRepository,
+    private readonly _notificationTypeRepository: NotificationTypeRepository,
+    private readonly _notificationRepository: NotificationRepository,
+    private readonly _socketManagementService: SocketManagementService,
+    private readonly _shareResultRequestService: ShareResultRequestService,
+    private readonly _userRepository: UserRepository,
+  ) {}
+
+  async emitResultNotification(
+    notificationLevel: NotificationLevelEnum,
+    notificationType: NotificationTypeEnum,
+    userIds: number[],
+    emmiterUser: number,
+    resultId: number,
+  ) {
+    try {
+      const notificationLevelData =
+        await this._notificationLevelRepository.findOne({
+          where: { type: notificationLevel },
+        });
+
+      const notificationTypeData =
+        await this._notificationTypeRepository.findOne({
+          where: { type: notificationType },
+        });
+
+      const filteredUserIds = userIds.filter(
+        (userId) => userId !== emmiterUser,
+      );
+
+      for (const userId of filteredUserIds) {
+        await this._notificationRepository.save({
+          target_user: userId,
+          emitter_user: emmiterUser,
+          result_id: resultId,
+          notification_level: notificationLevelData.notifications_level_id,
+          notification_type: notificationTypeData.notifications_type_id,
+        });
+      }
+
+      const usersOnline = await this._socketManagementService.getActiveUsers();
+      const usersOnlineIds = usersOnline.response.map(
+        (user: { userId: number }) => user.userId,
+      );
+      const matchUsers = filteredUserIds.filter((userId) =>
+        usersOnlineIds.includes(userId),
+      );
+
+      if (matchUsers.length === 0) {
+        this._logger.warn('No online users to notify.');
+        return {
+          response: null,
+          message: 'No online users to notify.',
+          status: HttpStatus.OK,
+        };
+      }
+
+      const resultData = await this._notificationRepository.find({
+        select: this.getNotificattionSelect(),
+        relations: this.getNotificationRelations(),
+        where: {
+          result_id: resultId,
+          emitter_user: emmiterUser,
+          notification_level: notificationLevelData.notifications_level_id,
+          notification_type: notificationTypeData.notifications_type_id,
+        },
+        take: 1,
+      });
```

Comment:
@JuanGuzman-io if you only need 1 notification, please consider using `repository.findOne()` vs `repository.find({take: 1})`, as is clearer and can even have better performance